File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l = 0;
        int length = 0;
        int[] freq = new int[128];
        int n = s.length();

        for(int right=0;right<n;right++){
            char currentChar = s.charAt(right);
            freq[currentChar]++;

            while(freq[currentChar] > 1){
                freq[s.charAt(l)]--;
                l++;
            }

            length = Math.max(length,right-l+1);
        }
        return length;

    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        d = set()
        
        maxs = 0
        left = 0
        
        for right in range(len(s)):
            
            while s[right] in d:
                
                d.remove(s[left])
                left += 1
            
            d.add(s[right])
            
            maxs = max(maxs,right - left + 1)
        
        return maxs